Checked arithmetic by default. (#9465)
ethereum/solidity: docs/security-considerations.rst

=========================================

As in many programming languages, Solidity's integer types are not actually integers.

> They resemble integers when the values are small, but cannot represent arbitrarily large numbers.

Reviewer: ```suggestion
They resemble integers when the values are small, but do not represent arbitrary precision integers.
```

Author: Isn't precision something else?

The following code causes an overflow because the result of the addition is too large
to be stored in the type ``uint8``:

::

  uint8 x = 255;
  uint8 y = 1;
  return x + y;

> Solidity has two modes in which it deals with these overflows: Checked and non-checked or "wrapping" mode.

Reviewer: ```suggestion
Solidity has two modes in which it deals with these overflows: Checked and Unchecked or "wrapping" mode.
```

The default checked mode will detect overflows and cause a failing assertion. You can disable this check
using ``unchecked { ... }``, causing the overflow to be silently ignored. The above code would return
``0`` if wrapped in ``unchecked { ... }``.

Even in checked mode, please do not rely on these checks, since there are situations where an overflow

> can lead to a contract being stuck.

Reviewer: What does it mean by that? Is it true that solidity checked arithmetic is a replacement for safemath libraries? Do we want to add it to the docs?

Author: What do you mean by "that"? Yes, SafeMath should not be needed anymore.

Reviewer: > Even in checked mode, please do not rely on these checks, since there are situations where an overflow
can lead to a contract being stuck.

Can you give an example of this situation?

I also think it's a good idea to write somewhere that safe math libraries are redundant after 0.8.0.
